### PR TITLE
fix: off by one error in domain seperation

### PR DESCRIPTION
**Pull request by bgorlick**
Looked like  there was a fixed length assumption for version bytes causing off-by-one errors with an inaccurate tag length calculation potentially leading to hash collisions. The static domain tag construction is not reflecting actual content variations.

Hopefully this addresses it:
We can dynamically construct the domain tags based on actual content, ensuring uniqueness. This should accurately calculate full the tag length to prevent inconsistencies, and then from there we can update the digest with exact length bytes and then the constructed tag. My thoughts is it's safer to converts version and label bytes to strings safely using String::from_utf8. this version implements expect for error handling to clearly identify conversion failures. I looked over the broader class and it looks like the approach is to panic, so I kept it consistent instead of returning String, which I had initially considered. 

Tested with build and cargo bench, and everything passes.

### CODE REVIEW

**Comment by SWvheerden** [comment]
Thanks for the PR, 
Can you please run fmt by running:
```
Run cargo +nightly-2024-02-01 fmt --all -- --check
```

File: src/hashing.rs [hunk 1]
```diff
@@ -78,26 +78,48 @@ pub trait DomainSeparation {
...
+
+        let version_string = String::from_utf8(version[version_offset..].to_vec())
+            .expect("Version conversion failed: invalid UTF-8");
+        let label_string = if label.is_empty() {
+            String::new()
         } else {
-            // 3 additional bytes are 2 x '.' delimiters and 'v' tag for version
-            domain.len() + (3 - version_offset) + label.len() + 3
+            String::from_utf8(label.to_vec())
+                .expect("Label conversion failed: invalid UTF-8")
```
Comment (SWvheerden) [inline src/hashing.rs:109]: I am not a big fan of this approach. 
`S` is a str, now we convert it to bytes, only to convert it back to a string, then down below, convert the whole string back to bytes again?

Comment (bgorlick) [reply src/hashing.rs:109]: You raise a valid point, I will share a variation. My main mo was to get something initially to address the issue. Will refactor.

**Comment by AaronFeickert** [comment]
Can you provide an example of how the off-by-one error occurs? It's not immediately clear to me.

**Comment by AaronFeickert** [comment]
Superseded by #229. Closing.

**Pull request closed without merging**